### PR TITLE
Make ParametrizedEvolution a pytree

### PR DESCRIPTION
**Pull request by Qottmann**
Registering `ParametrizedEvolution` as a pytree in order to enable https://github.com/PennyLaneAI/pennylane/pull/3859

### CODE REVIEW

File: pennylane/pulse/parametrized_evolution.py [hunk 1]
```diff
@@ -570,6 +571,21 @@ def _format_arraylike(x):
...
+        hyper_params = self.hyper_parameters
+        odeint_kwargs = self.odeint_kwargs
+        
+        return (hamiltonian, params, t, hyper_params, odeint_kwargs)
+    
+    @classmethod
+    def _unflatten(cls, hamiltonian, params, t, hyper_params, odeint_kwargs):
+        return cls(hamiltonian, params, t, complementary=hyper_params["complementary"], return_intermediate=hyper_params["return_intermediate"], **odeint_kwargs)
+    
+register_pytree(ParametrizedEvolution, ParametrizedEvolution._flatten, ParametrizedEvolution._unflatten)
```
[the remark this answers is not in the document]
Comment (Qottmann) [reply pennylane/pulse/parametrized_evolution.py]: I learned that all subclasses of operation are registered as pytrees (also `ParametrizedEvolution`, but the default flatten and unflatten functions dont work) https://github.com/PennyLaneAI/pennylane/blob/ac0a6b7246e2171c80b64996a2e29a7ea71c9485/pennylane/operation.py#L689

In regards to testing explicitly for pytree registration, good question, I've seen some tests that check the `_flatten` and `_unflatten` functions. Checking explicitly that it is indeed a registered pytree I haven't seen, except for this one but I think this is rather exotic: https://github.com/PennyLaneAI/pennylane/blob/master/tests/ops/test_identity.py#L37

File: pennylane/pulse/parametrized_evolution.py [hunk 2]
```diff
@@ -472,6 +474,53 @@ def hash(self):
...
+            [
+                str(self.name) == str(other.name),
+                self.wires == other.wires,
+                self.hyperparameters == other.hyperparameters,
+                self.H == other.H,
+                self.odeint_kwargs == other.odeint_kwargs,
+                same_times,
+                same_data,
+            ]
+        )
```
Comment (Qottmann) [inline pennylane/pulse/parametrized_evolution.py:522]: This turned out longer and uglier than anticipated, happy to hear suggestions to make it more concise

File: tests/pulse/test_parametrized_evolution.py [hunk 3]
```diff
@@ -36,8 +36,48 @@ class MyOp(qml.RX):  # pylint: disable=too-few-public-methods
...
+
+    @pytest.mark.parametrize("evol", example_pytree_evolutions)
+    def test_flatten_unflatten_identity(self, evol):
+        """Test that flattening and unflattening is yielding the same parametrized evolution"""
+        assert evol._unflatten(*evol._flatten()) == evol
+
+
 @pytest.mark.jax
-@pytest.mark.xfail  # TODO: fix by story 50356
+@pytest.mark.xfail
```
Comment (Qottmann) [inline tests/pulse/test_parametrized_evolution.py:80]: Still cant quite get this to work because there is some non-hashable metadata, i.e. the hyperparamters (dictionary) and odeint_kwargs (dictionary) 🤔

File: pennylane/pulse/parametrized_evolution.py [hunk 4]
```diff
@@ -472,6 +474,53 @@ def hash(self):
             )
         )
 
+    def _flatten(self):
+        """Unpacking the ParametrizedEvolution pytree into a flat tuple"""
+
+        data = self.data
+        metadata = (tuple(self.t), self.H, self.hyperparameters, self.odeint_kwargs)
```
Comment (Qottmann) [inline pennylane/pulse/parametrized_evolution.py:481]: How can I make hyperparameters and odeint_kwargs hashable?

**Comment by Qottmann** [comment]
There are some problems arising from making `params` and `t` optional keyword arguments. I dont see a reason why they should be, an evolution with a time or parameters doesnt make much sense. Was there any programmatic reasons to do that, that I currently dont recall? @mudit2812 @lillian542 @dwierichs

**Comment by lillian542** [comment]
> There are some problems arising from making `params` and `t` optional keyword arguments of `ParametrizedEvolution`. I dont see a reason why they should be, an evolution with a time or parameters doesnt make much sense. Was there any programmatic reasons to do that, that I currently dont recall? @mudit2812 @lillian542 @dwierichs

IIRC it was more of a UI thing. It's what allows us to do this:

```
op = qml.evolve(Ht)

res_t1 = op(params, t1)
res_t2 = op(params, t2)
```

instead of 

```
res_t1 = qml.evolve(Ht, params, t1)
res_t2 = qml.evolve(Ht, params, t2)
```

If the question is why they are not required in the `__init__`, the idea was to have the parameterized evolution initialize _without_ parameters, and then update it with parameters to get the result at particular spots in parameter space. Similar to how a `ParameterizedHamiltonian` is not fully defined until you call it with parameters, but the idea was to initialize without.

If the question is why they are in the initialization at all, and not only in the `__call__` like for `ParameterizedHamiltonian`, I don't recall the initial reason, and we talked about removing it at one point. But we use it as a developer-facing way to create the operators in the AWS Braket tests without introducing a `jax` dependency to the AWS-Braket test suite, so we couldn't. But in the meeting where we settled that I believe we decided the UI where you pass them to the `__init__` would not be promoted anywhere in the docs, and was a developer-facing feature for hardware tests.

Tl;dr: to my knowledge no programmatic reason, but a UI choice that we've used consistently in the docs and demos.

File: pennylane/pulse/parametrized_evolution.py (hunk 2 above)
Comment (mudit2812) [reply pennylane/pulse/parametrized_evolution.py:520]: Is `qml.equal` not enough to use with `ParametrizedEvolution.__eq__`? If not, maybe it's meaningful to update `qml.equal`.

File: pennylane/pulse/parametrized_evolution.py (hunk 4 above)
Comment (lillian542) [reply pennylane/pulse/parametrized_evolution.py:479]: Since we know what the only two things in hyperparameters are, could we handle them like this?

```suggestion
        metadata = (tuple(self.t), self.H, self.hyperparameters["complementary"], self.hyperparameters["return_intermediate"], self.odeint_kwargs)
```

and then in `_unflatten` we pass `metadata[2]` to `complementary` and `metadata[3]` to `return_intermediate`

Comment (lillian542) [reply pennylane/pulse/parametrized_evolution.py:479]: For `odeint_kwargs` its a little harder since we don't know how many or what they will be called

Comment (lillian542) [reply pennylane/pulse/parametrized_evolution.py:479]: How about something like this:

```
sorted_odeint_kwargs = sorted(op.odeint_kwargs.items())
odeint_kwargs = tuple([kwarg for kwarg, val in sorted_odeint_kwargs] + [val for val, val in sorted_odeint_kwargs])
```
to create a hashable tuple with all the info, and then in `_unflatten`:

```
idx = int(len(odeint_kwargs)/2)
keys = list(odeint_kwargs)[:idx]
vals = list(odeint_kwargs)[idx:]
odeint_kwarg_dict = {k: v for k, v in zip(keys, vals)}
odeint_kwarg_dict
```
and pass that to the operator init?

File: pennylane/pulse/parametrized_evolution.py (hunk 2 above)
Comment (Qottmann) [reply pennylane/pulse/parametrized_evolution.py:520]: To be more concise, this is coming from the `None` data in `self.t`.
```python
>>> np.isclose(None, None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/Users/korbinian.kottmann/Xanadu/Sandbox2/sandbox2.ipynb Cell 2 line 1
----> 1 np.isclose(None, None)

File ~/anaconda3/envs/pennylane/lib/python3.11/site-packages/numpy/core/numeric.py:2348, in isclose(a, b, rtol, atol, equal_nan)
   2345     dt = multiarray.result_type(y, 1.)
   2346     y = asanyarray(y, dtype=dt)
-> 2348 xfin = isfinite(x)
   2349 yfin = isfinite(y)
   2350 if all(xfin) and all(yfin):

TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

File: pennylane/pulse/parametrized_evolution.py [hunk 5]
```diff
@@ -472,6 +473,52 @@ def hash(self):
...
+            **odeint_kwargs,
+        )
+
+    def __eq__(self, other):
+        """Check equality between two ParametrizedEvolution operators"""
+        if self.t is None:
+            same_times = other.t is None
+        else:
+            same_times = all(self.t == other.t)
+        if self.data is None:
```
Comment (lillian542) [inline pennylane/pulse/parametrized_evolution.py:502]: is `self.data` ever `None`? It looks like its `()` if there is no data

File: pennylane/pulse/parametrized_evolution.py [hunk 6]
```diff
@@ -472,6 +473,52 @@ def hash(self):
...
+            same_data = other.data is None
+        elif isinstance(self.data, tuple) and isinstance(other.data, tuple):
+            same_data = self.data == other.data
+        elif isinstance(self.data, Iterable) and isinstance(other.data, Iterable):
```
Comment (lillian542) [inline pennylane/pulse/parametrized_evolution.py:506]: Under what circumstance does this happen? I thought we pass params to the `Operation` init and it always converts to a tuple. I think the tuple comparison in 504-505 should be sufficient for comparing `data` attributes.

File: pennylane/pulse/parametrized_evolution.py [hunk 7]
```diff
@@ -472,6 +473,52 @@ def hash(self):
...
+        return cls(
+            H,
+            params,
+            t,
+            complementary=hyperparamameters["complementary"],
+            return_intermediate=hyperparamameters["return_intermediate"],
+            **odeint_kwargs,
+        )
+
+    def __eq__(self, other):
```
Comment (lillian542) [inline pennylane/pulse/parametrized_evolution.py:496]: Why do we need this method defined here? It already inherits

```
    def __eq__(self, other):
        return qml.equal(self, other)
```

from `Operator`, and we have an equality method defined for `ParameterizedEvolution` there. If there are aspects missing there, we should update `ops.functions.equal.py`.

Comment (lillian542) [reply pennylane/pulse/parametrized_evolution.py:496]: Ah, I can see @mudit2812 already asked

File: pennylane/pulse/parametrized_evolution.py (hunk 2 above)
Comment (lillian542) [reply pennylane/pulse/parametrized_evolution.py:520]: Could we just update the logic regarding `op.t` to account for the `None` case in the standard `equal` function then for `ParameterizedEvolution` then?

File: tests/pulse/test_parametrized_evolution.py [hunk 8]
```diff
@@ -36,8 +36,48 @@ class MyOp(qml.RX):  # pylint: disable=too-few-public-methods
     has_diagonalizing_gates = False
 
 
+def amp(p, t):
+    return p * t
+
+
+H = qml.PauliX(1) + amp * qml.PauliZ(0) + amp * qml.PauliY(1)
```
Comment (lillian542) [inline tests/pulse/test_parametrized_evolution.py:43]: Could we also add a Hamiltonian with a ragged param shape? Something like one coefficient has 1 parameter and the other has 2.

File: tests/pulse/test_parametrized_evolution.py [hunk 9]
```diff
@@ -36,8 +36,48 @@ class MyOp(qml.RX):  # pylint: disable=too-few-public-methods
...
+params = [0.5, 0.5]
+
+example_pytree_evolutions = [
+    qml.pulse.ParametrizedEvolution(H, params, t=0.5),
```
Comment (lillian542) [inline tests/pulse/test_parametrized_evolution.py:47]: Do you want to also test one that is initialized but without `params` and `t`? Since it came up in working with the `equal` function.

Comment (Qottmann) [reply tests/pulse/test_parametrized_evolution.py:48]: Yep! ✅

File: pennylane/pulse/parametrized_evolution.py (hunk 4 above)
Comment (Qottmann) [reply pennylane/pulse/parametrized_evolution.py]: Used a solution Christina suggested on slack that just creates a tuple of tuples from dict.items()

File: pennylane/pulse/parametrized_evolution.py (hunk 2 above)
Comment (Qottmann) [reply pennylane/pulse/parametrized_evolution.py]: Workaround that doesnt seem to hurt anything: trivially initialize the time as `(0., 0.)`. Any objections?

File: tests/pulse/test_parametrized_evolution.py [hunk 10]
```diff
@@ -36,18 +36,82 @@ class MyOp(qml.RX):  # pylint: disable=too-few-public-methods
...
+    """Testing pytree related functionality"""
+
+    @pytest.mark.parametrize("evol", example_pytree_evolutions)
+    def test_flatten_unflatten_identity(self, evol):
+        """Test that flattening and unflattening is yielding the same parametrized evolution"""
+        assert evol._unflatten(*evol._flatten()) == evol
+
+
 @pytest.mark.jax
-@pytest.mark.xfail  # TODO: fix by story 50356
```
Comment (Qottmann) [inline tests/pulse/test_parametrized_evolution.py:40]: When I run the test in a python shell it works, but in pytest I get the following error. What is this pickling business about?

```pycon
pennylane/ops/functions/assert_valid.py:284: in assert_valid
    _check_pickle(op)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

op = ParametrizedEvolution(0.5, wires=[0])

    def _check_pickle(op):
        """Check that an operation can be dumped and reloaded with pickle."""
>       pickled = pickle.dumps(op)
E       AttributeError: Can't pickle local object 'test_standard_validity.<locals>.f1'

pennylane/ops/functions/assert_valid.py:204: AttributeError
```

**Comment by Qottmann** [comment]
Thanks for the great input! I went with the workaround to initialize `self.t = [0., 0.]` in case `None` is passed. This doesnt seem to hurt any tests (other than those that explicitly test for that assumption) and we can use standard equality checks.
The hashable metadata problem could also be resolved, there is quite a general way to pack and unpack dictionaries: `tuple((key, value) for key, value in my_dict.items())`.

The only thing left is the validy test raises an error about not being able to pickle some properties. When I run just the test somewhere it works, but in pytest it doesnt. I see this test in many other functions in the codebase so something must be off here, any idea what?

File: tests/pulse/test_parametrized_evolution.py (hunk 10 above)
Comment (albi3ro) [reply tests/pulse/test_parametrized_evolution.py:40]: So `f1` is a local function, so its not pickleable.  

Since we know that pickling in finicky and a lot of things can cause it to fail, you can just do `assert_valid(ev, skip_pickle=True)`.  

Or you can move the function out of the function to be global.

File: pennylane/pulse/parametrized_evolution.py (hunk 2 above)
Comment (lillian542) [reply pennylane/pulse/parametrized_evolution.py]: None from me, duration 0 seems like the same thing as `t=None`!

File: tests/pulse/test_parametrized_evolution.py (hunk 10 above)
Comment (Qottmann) [reply tests/pulse/test_parametrized_evolution.py:40]: Thanks!

**Comment by lillian542** [comment]
I'm slightly partial to leaving `t=None` then, and just updating `qml.equal` with something like the logic you initially suggested, i.e.

```
    # check times match
    if None in [op1.t, op2.t]:
        if op1.t != op2.t:
            return False
    elif not qml.math.allclose(op1.t, op2.t):
        return False
```

instead of the current logic for comparing the times. But I'm also open to just removing those checks 🤷‍♀️

**Comment by Qottmann** [comment]
I went back to the default `evol.t = None`, much cleaner like this. Thanks for insisting @lillian542 , I hadnt realized the there is a separte dispatch for `qml.equal` in the case of `ParametrizedEvolution`.

File: pennylane/pulse/parametrized_evolution.py [hunk 11]
```diff
@@ -472,6 +472,39 @@ def hash(self):
...
+    def _unflatten(cls, data, metadata):
+        if len(data) == 0:
+            params = None
+        else:
+            params = data
+        t, H, return_intermediate, complementary, dense, odeint_kwargs = metadata
+
+        return cls(
+            H,
+            params,
```
Comment (albi3ro) [inline pennylane/pulse/parametrized_evolution.py:500]:
```suggestion
            None if len(data) == 0 else data,
```

**Review by lillian542 — APPROVED** [review]
Nice! This looks good to me 🎉
